File: src/java/3_StructuralPatterns/3_Composite/GeometricShapes.java
package com.activemesa.structural.composite;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class GraphicObject
{
  protected String name = "Group";

  public String getName()
  {
    return name;
  }

  public void setName(String name)
  {
    this.name = name;
  }

  public GraphicObject()
  {
  }

  public String color;
  public List<GraphicObject> children = new ArrayList<>();

  private void print(StringBuilder stringBuilder,  int depth)
  {
    stringBuilder.append(String.join("", Collections.nCopies(depth, "*")))
      .append(depth > 0 ? " " : "")
      .append((color == null || color.isEmpty()) ? "" : color + " ")
      .append(getName())
      .append(System.lineSeparator());
    for (GraphicObject child : children)
      child.print(stringBuilder,  depth+1);
  }

  @Override
  public String toString()
  {
    StringBuilder sb = new StringBuilder();
    print(sb, 0);
    return sb.toString();
  }
}

class Circle extends GraphicObject
{
  public Circle(String color)
  {
    name = "Circle";
    this.color = color;
  }
}

class Square extends GraphicObject
{
  public Square(String color)
  {
    name = "Square";
    this.color = color;
  }
}

class GeometricShapesDemo
{
  public static void main(String[] args)
  {
    GraphicObject drawing = new GraphicObject();
    drawing.setName("My Drawing");
    drawing.children.add(new Square("Red"));
    drawing.children.add(new Circle("Yellow"));

    GraphicObject group = new GraphicObject();
    group.children.add(new Circle("Blue"));
    group.children.add(new Square("Blue"));
    drawing.children.add(group);

    System.out.println(drawing);
  }
}

File: src/csharp/3_StructuralPatterns/3_Composite/GeometricShapes.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using static System.Console;

namespace DotNetDesignPatternDemos.Structural.Composite.GeometricShapes
{
  public class GraphicObject
  {
    public virtual string Name { get; set; } = "Group";
    public string Color;
    private Lazy<List<GraphicObject>> children = new Lazy<List<GraphicObject>>();
    public List<GraphicObject> Children => children.Value;

    private void Print(StringBuilder sb, int depth)
    {
      sb.Append(new string('*', depth))
        .Append(string.IsNullOrWhiteSpace(Color) ? string.Empty : $"{Color} ")
        .AppendLine($"{Name}");
      foreach (var child in Children)
        child.Print(sb, depth + 1);
    }

    public override string ToString()
    {
      var sb = new StringBuilder();
      Print(sb, 0);
      return sb.ToString();
    }
  }
  
  public class Circle : GraphicObject
  {
    public override string Name => "Circle";
  }

  public class Square : GraphicObject
  {
    public override string Name => "Square";
  }

  public class Demo
  {
    static void Main(string[] args)
    {
      var drawing = new GraphicObject {Name = "My Drawing"};
      drawing.Children.Add(new Square {Color = "Red"});
      drawing.Children.Add(new Circle{Color="Yellow"});
      
      var group = new GraphicObject();
      group.Children.Add(new Circle{Color="Blue"});
      group.Children.Add(new Square{Color="Blue"});
      drawing.Children.Add(group);

      WriteLine(drawing);
    }
  }
}
